synapse/handlers/message.py

```diff
@@ -51,25 +61,70 @@ def __init__(self, hs):
 
         self.pagination_lock = ReadWriteLock()
         self._purges_in_progress_by_room = set()
+        # map from purge id to PurgeStatus
+        self._purges_by_id = {}
 
-    @defer.inlineCallbacks
-    def purge_history(self, room_id, topological_ordering,
-                      delete_local_events=False):
+    def start_purge_history(self, room_id, topological_ordering,
+                            delete_local_events=False):
         if room_id in self._purges_in_progress_by_room:
             raise SynapseError(
                 400,
                 "History purge already in progress for %s" % (room_id, ),
             )
 
+        purge_id = random_string(16)
+        self._purges_by_id[purge_id] = PurgeStatus()
+        run_in_background(
+            self._purge_history,
+            purge_id, room_id, topological_ordering, delete_local_events,
+        )
+        return purge_id
+
+    @defer.inlineCallbacks
+    def _purge_history(self, purge_id, room_id, topological_ordering,
+                       delete_local_events):
+        """Carry out a history purge on a room.
+
+        Args:
+            purge_id (str): The id for this purge
+            room_id (str): The room to purge from
+            topological_ordering (int): minimum topo ordering to preserve
+            delete_local_events (bool): True to delete local events as well as
+                remote ones
+
+        Returns:
+            Deferred
+        """
         self._purges_in_progress_by_room.add(room_id)
         try:
             with (yield self.pagination_lock.write(room_id)):
                 yield self.store.purge_history(
                     room_id, topological_ordering, delete_local_events,
                 )
+            logger.info("[purge] complete")
+            self._purges_by_id[purge_id].status = "complete"
+        except Exception:
+            logger.error("[purge] failed: %s", Failure().getTraceback().rstrip())
+            self._purges_by_id[purge_id].status = "failed"
         finally:
             self._purges_in_progress_by_room.discard(room_id)
 
+            # remove the purge from the list an hour after it completes
+            def clear_purge():
+                del self._purges_by_id[purge_id]
```

Comment:
For paranoia's sake I tend to prefer `d.pop(key, None)`, as that won't throw if the key doesn't exist, but shrug.

Reply:
I'd argue that knowing the key has gone missing would be a good thing...

Reply:
fair